File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> charSet = new HashSet<>();
        int len = s.length();
        int left = 0;
        int right = 0;
        int maxLength = 0;
        
        while (right < len) {
            char currentCharacter = s.charAt(right);
            if (charSet.contains(currentCharacter)) {
                while (left <= right) {
                    char tempCharacter = s.charAt(left);
                    left++;
                    if (currentCharacter == tempCharacter) {
                        break;
                    } else {
                        charSet.remove(tempCharacter);
                    }
                }
            } else {
                charSet.add(currentCharacter);
            }
            right++;
            maxLength = Math.max(maxLength, right - left);
        }
        return maxLength;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
from collections import Counter
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        bucket = Counter()
        n = len(s)
        ans = 0
        left = 0

        for right in range(n):
            bucket[s[right]] += 1

            while bucket[s[right]] > 1:
                bucket[s[left]] -= 1
                left += 1

            ans = max(ans, right - left + 1)

        return ans